source/particle/world.cc

```diff
@@ -0,0 +1,889 @@
+                                                     *  std::pow(2,dim);
+          data_offset = triangulation.register_data_attach(transfer_size_per_cell,callback_function);
+        }
+    }
+
+    template <int dim>
+    void
+    World<dim>::register_load_callback_function(typename parallel::distributed::Triangulation<dim> &triangulation)
+    {
+      Assert(particles.size() == 0,
+             ExcMessage("We are in the process of mesh refinement. All tracers "
+                        "should have been serialized and stored, but there are still some "
+                        "around. Is there a bug in the storage function?"));
+
+      // Check if something was stored
+      if (data_offset != numbers::invalid_unsigned_int)
+        {
+          const std_cxx11::function<void(const typename parallel::distributed::Triangulation<dim>::cell_iterator &,
+                                         const typename parallel::distributed::Triangulation<dim>::CellStatus,
+                                         const void *) > callback_function
+            = std_cxx11::bind(&aspect::Particle::World<dim>::load_tracers,
+                              std_cxx11::ref(*this),
+                              std_cxx11::_1,
+                              std_cxx11::_2,
+                              std_cxx11::_3);
+
+          triangulation.notify_ready_to_unpack(data_offset,callback_function);
+          data_offset = numbers::invalid_unsigned_int;
+        }
+    }
+
+    template <int dim>
+    unsigned int
+    World<dim>::cell_weight(const typename parallel::distributed::Triangulation<dim>::cell_iterator &cell,
+                            const typename parallel::distributed::Triangulation<dim>::CellStatus status)
+    {
+      if (cell->active() && !cell->is_locally_owned())
+        return 0;
+
+      if (status == parallel::distributed::Triangulation<dim>::CELL_PERSIST
+          || status == parallel::distributed::Triangulation<dim>::CELL_REFINE)
+        {
+          const types::LevelInd found_cell = std::make_pair<int, int> (cell->level(),cell->index());
+          const unsigned int n_particles_in_cell = particles.count(found_cell);
+          return n_particles_in_cell * tracer_weight;
+        }
+      else if (status == parallel::distributed::Triangulation<dim>::CELL_COARSEN)
+        {
+          unsigned int n_particles_in_cell = 0;
+
+          for (unsigned int child_index = 0; child_index < cell->number_of_children(); ++child_index)
+            {
+              const typename parallel::distributed::Triangulation<dim>::cell_iterator child = cell->child(child_index);
+              const types::LevelInd found_cell = std::make_pair<int, int> (child->level(),child->index());
+              n_particles_in_cell += particles.count(found_cell);
+            }
+          return n_particles_in_cell * tracer_weight;
+        }
+
+      return 0;
```

Comment:
is there a case where we could ever get here? Otherwise maybe add an `Assert (false, ExcInternalError())`?